cogeo_mosaic/backend/s3.py

```diff
@@ -0,0 +1,51 @@
+import functools
+from typing import BinaryIO, Dict
+
+import mercantile
+from boto3.session import Session as boto3_session
+
+from cogeo_mosaic.backend.base import BaseBackend
+from cogeo_mosaic.backend.utils import get_assets_from_json
+from cogeo_mosaic.utils import _decompress_gz
+
+
+class S3Backend(BaseBackend):
+    """S3 Backend Adapter"""
+
+    def __init__(self, bucket: str, key: str):
```

Comment:
can we assume we are passing a full s3 path here ? and maybe an optional client ?

Reply:
> can we assume we are passing a full s3 path here

I think yes... If we assume the url is of type
```
s3://<bucket>/<key>
```
then we always have the full path to the bucket.

If we maintain the current `/mosaicid/{endpoint}` then we can add the bucket prefix there before passing to `S3Backend`

> and maybe an optional client

I added `client` instantiation to `__init__`, so you'll be able to do
```py
with S3Backend(bucket, key) as mosaic:
	s3client = mosaic.client
```

Reply:
well but idealy when we do
```
with S3Backend(bucket, key) as mosaic:
```
the __enter__ method will fetch the mosaic using a client 

that why I think doing 
```
with S3Backend(s3path, client=myclient) as mosaic:
```
would be better

Reply:
> If we maintain the current /mosaicid/{endpoint} then we can add the bucket prefix there before passing to S3Backend` 

The mosaicid based endpoint should be considered as `custom` and maybe removed ? ir maybe we add an environment variable telling which is our fall back method (s3 or dynamo)...

Reply:
`__init__` is called on `__enter__` I believe. Putting logic in `__init__` and having `__enter__` return just `self` looks to be common:
https://stackoverflow.com/a/34749985

Reply:
maybe I'm not exlaining mysefl right. 

the idea is to define the s3 client outside the backend 

```
from boto3.session import Session as boto3_session
session = boto3_session()
client = session.client("s3")

with S3Backend(s3path, client=myclient) as mosaic:
    assets = mosaic.tile(1,1,1)
```

Reply:
Ok, yeah, I added an optional `client` arg here:
https://github.com/developmentseed/cogeo-mosaic/pull/38/files#diff-998203b2b4af94384b3c4a81c56c57f5R19

Maybe I should do that for `dynamodb` too... though I'm using `boto3.resource` for it, not `boto3.Session.client`... not sure what the difference is

Reply:
cool, 
they are mostly the same, just different ways to call them. I problem boto3_session because I think you could pass more options to it, but that's maybe not true

Reply:
My `dynamodb` code only seems to work with `boto3.resource`, not `boto3.Session().client()` so I'll leave it for now

Reply:
cool